staging/vhost-device-sound/src/audio_backends/pipewire/test_utils.rs

```diff
@@ -0,0 +1,134 @@
+// Manos Pitsidianakis <manos.pitsidianakis@linaro.org>
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use std::{
+    io::Read,
+    path::Path,
+    process::{Child, Command, Stdio},
+};
+
+use tempfile::{tempdir, TempDir};
+
+/// Temporary Dbus session which is killed in drop().
+pub struct DbusSession {
+    pub child: Child,
+    pub address: String,
+}
+
+impl DbusSession {
+    pub fn new(working_dir: &Path) -> Self {
+        let address_prefix = format!("unix:path={}", working_dir.join("dbus").display());
+        let child = Command::new("/usr/bin/dbus-daemon")
+            .args(["--session", "--address", &address_prefix, "--print-address"])
+            .env("DBUS_VERBOSE", "1")
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .stdin(Stdio::null())
+            .current_dir(working_dir)
+            .spawn()
+            .expect("ERROR: dbus-daemon binary not found");
+
+        Self {
+            child,
+            address: address_prefix,
+        }
+    }
+}
+
+impl Drop for DbusSession {
+    fn drop(&mut self) {
+        println!("INFO: Killing Dbus session {}", self.child.id());
+        if let Err(err) = self.child.kill() {
+            println!(
+                "ERROR: could not kill dbus process {}: {err}",
+                self.child.id()
+            );
+        }
+        // We mustn't panic in drop(), so use a wrapper function for convenience
+        print_output(&mut self.child, "dbus");
+    }
+}
+
+/// The pipewire test harness. It must only be constructed via
+/// `PipewireTestHarness::new()`.
+#[non_exhaustive]
+pub struct PipewireTestHarness {
+    pub dbus: DbusSession,
+    pub pipewire_child: Child,
+    pub tempdir: TempDir,
+}
+
+pub fn launch_pipewire(
+    tempdir: &Path,
+    dbus_session_bus_address: &Path,
+) -> Result<Child, std::io::Error> {
+    Command::new("pipewire")
+        .env("DBUS_SESSION_BUS_ADDRESS", dbus_session_bus_address)
+        .env("XDG_RUNTIME_DIR", tempdir)
+        .current_dir(tempdir)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .stdin(Stdio::null())
+        .spawn()
+}
+
+impl PipewireTestHarness {
+    pub fn new() -> Self {
+        let tempdir = tempdir().unwrap();
+
+        let dbus_session = DbusSession::new(tempdir.path());
+        println!("INFO: dbus_session_bus_address={}", dbus_session.address);
+
+        println!("INFO: Wait for dbus to setup...");
+        std::thread::sleep(std::time::Duration::from_secs(1));
```

Comment:
It sound like waiting for the address to get printed by the daemon is recommended to wait until the daemon is fully up [1]. That might be a more reliable way than waiting a hardcoded second? :thinking:

[1] https://dbus.freedesktop.org/doc/dbus-daemon.1.html
```
On Unix platforms, a similar result can be achieved by waiting for the address and/or process ID to be printed to the inherited file descriptors used for --print-address and/or --print-pid.
````

Reply:
That was the original code, this works just as fine.

Reply:
Not a huge fan, but I guess if it creates problems later we can fix them then.